src/client.rs

```diff
@@ -359,35 +363,58 @@ impl Future for ResponseFuture {
 
 // ===== impl Peer =====
 
-impl proto::Peer for Peer {
-    type Send = Request<()>;
-    type Poll = Response<()>;
-
-
-    fn dyn() -> proto::DynPeer {
-        proto::DynPeer::Client
-    }
-
-    fn is_server() -> bool {
-        false
-    }
-
-    fn convert_send_message(id: StreamId, request: Self::Send, end_of_stream: bool) -> Headers {
+impl Peer {
+    pub fn convert_send_message(
+        id: StreamId,
+        request: Request<()>,
+        end_of_stream: bool) -> Result<Headers, SendError>
+    {
         use http::request::Parts;
 
         let (
             Parts {
                 method,
                 uri,
                 headers,
+                version,
                 ..
             },
             _,
         ) = request.into_parts();
 
+        let is_connect = method == Method::CONNECT;
+
         // Build the set pseudo header set. All requests will include `method`
         // and `path`.
-        let pseudo = Pseudo::request(method, uri);
+        let mut pseudo = Pseudo::request(method, uri);
+
+        if pseudo.scheme.is_none() {
+            // If the scheme is not set, then there are a two options.
+            //
+            // 1) Authority is not set. In this case, a request was issued with
+            //    a relative URI. This is permitted **only** when forwarding
+            //    HTTP 1.x requests. If the HTTP version is set to 2.0, then
+            //    this is an error.
+            //
+            // 2) Authority is set, then the HTTP method *must* be CONNECT.
+            //
+            // It is not possible to have a scheme but not an authority set (the
+            // `http` crate does not allow it).
+            //
+            if pseudo.authority.is_none() {
+                if version == Version::HTTP_2 {
+                    return Err(UserError::MissingUriSchemeAndAuthority.into());
+                } else {
+                    // This is acceptable as per the above comment. However,
+                    // HTTP/2.0 requires that a scheme is set. Since we are
+                    // forwarding an HTTP 1.1 request, the scheme is set to
+                    // "http".
+                    pseudo.set_scheme(uri::Scheme::HTTP);
```

Comment:
I hate the `:scheme` header, it's so worthless. Should we be defaulting to `http` if it's not present? Ah what's it matter anyways...

Reply:
I am only defaulting it to "http" when dealing w/ HTTP/1.1 requests, which really have no way of setting the scheme w/o setting an authority (because it doesn't make sense to?)

If sending an HTTP/2.0 request w/o a scheme + authority, the lib will reject it.

1.1 vs. 2.0 is determined by the version field on the request struct.